### PR TITLE
[WIP] Prevent package uninstall when active CRs

### PR DESCRIPTION
**Pull request by turkenh**
### Description of your changes

Quick and dirty implementation to check whether there are active CRs for CRDs managed by a package before uninstalling it.

Fixes https://github.com/crossplane/crossplane/issues/4251

I have:

- [ ] Read and followed Crossplane's [contribution process].
- [ ] Added or updated unit **and** E2E tests for my change.
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR, if necessary.
- [ ] Opened a PR updating the [docs], if necessary.

[contribution process]: https://git.io/fj2m9
[docs]: https://docs.crossplane.io/contribute/contribute

### CODE REVIEW

File: cmd/crossplane/core/core.go [hunk 1]
```diff
@@ -190,7 +190,7 @@ func (c *startCommand) Run(s *runtime.Scheme, log logging.Logger) error { //noli
 		LeaseDuration:              func() *time.Duration { d := 60 * time.Second; return &d }(),
 		RenewDeadline:              func() *time.Duration { d := 50 * time.Second; return &d }(),
 
-		HealthProbeBindAddress: ":5000",
+		HealthProbeBindAddress: ":5001",
```
[the remark this answers is not in the document]
Comment (turkenh) [reply cmd/crossplane/core/core.go:193]: No, I didn't intend to commit. 
It conflicted with some service on my Mac when I want to run XP locally and hacked like this after noticing that I cannot stop that other service.

```
❯ lsof -ti:5000
45590
❯ ps -ef | grep 45590
501 45590     1   0  3:37PM ??         0:04.27 /System/Library/CoreServices/ControlCenter.app/Contents/MacOS/ControlCenter
```

File: internal/controller/pkg/manager/reconciler.go [hunk 2]
```diff
@@ -293,6 +314,67 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		return reconcile.Result{}, err
 	}
 
+	if meta.WasDeleted(p) {
+		if len(prs.GetRevisions()) > 0 {
```
Comment (sttts) [inline internal/controller/pkg/manager/reconciler.go:318]:
```suggestion
	if meta.WasDeleted(p) && len(prs.GetRevisions()) > 0 {
```

File: internal/controller/pkg/manager/reconciler.go [hunk 3]
```diff
@@ -293,6 +314,67 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
...
+			for _, pr := range prs.GetRevisions() {
+				if pr.GetDesiredState() == v1.PackageRevisionActive {
```
Comment (sttts) [inline internal/controller/pkg/manager/reconciler.go:320]:
```suggestion
				if pr.GetDesiredState() != v1.PackageRevisionActive {
					continue
				}
```

**Comment by turkenh** [comment]
The primary motivation for this PR "now" was the hope that it could help with the current e2e flakiness, but unfortunately, it didn't help as is.

While testing this, I observed the following two odd behaviors:

1. The test failed while waiting for the provider to be deleted.
- There is an MR without a deletion timestamp, but its owner (the composite) does not exist. Even the CRD for the composite is gone, but the GC does not clean it up.
- Hence, the provider properly rejects uninstallation.
- Not much related, but observing the following logs in Crossplane pods:
  ```shell
  W1016 14:18:19.380873       1 reflector.go:535] k8s.io/client-go@v0.28.2/tools/cache/reflector.go:229: failed to list nop.example.org/v1alpha1, Kind=XNopResource: xnopresources.nop.example.org is forbidden: User "system:serviceaccount:crossplane-system:crossplane" cannot list resource "xnopresources" in API group "nop.example.org" at the cluster scope
  E1016 14:18:19.381001       1 reflector.go:147] k8s.io/client-go@v0.28.2/tools/cache/reflector.go:229: Failed to watch nop.example.org/v1alpha1, Kind=XNopResource: failed to list nop.example.org/v1alpha1, Kind=XNopResource: xnopresources.nop.example.org is forbidden: User "system:serviceaccount:crossplane-system:crossplane" cannot list resource "xnopresources" in API group "nop.example.org" at the cluster scope
  ```

2. The test failed while waiting for the function package to be deleted.
- No XRs/MRs left, no provider package or providerrevision.
- Only a function package left, with deletion timestamp and no events related. I couldn't figure out why it is not deleted.
- Restarting Crossplane pod cleans up the function package without any other manual intervention.

**Pull request closed without merging**